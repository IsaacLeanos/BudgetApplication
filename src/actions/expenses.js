import uuid from 'uuid'
import database from '../firebase/firebase'


export const addExpense=(expense)=>{
    return{
        type:'ADD_EXPENSE',
        expense:expense
    }
}
export const startAddExpense=({description='',note='',amount=0,createdAt=0}={})=>{
    return(dispatch,getState)=>{
        const uid=getState().authen.uid
        const expense={description,note,amount,createdAt}
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        }).catch((e)=>{console.log(e)})
    }
}
export const startRemoveExpense=({id}={})=>{
    return(dispatch,getState)=>{
        const uid=getState().authen.uid
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id:id}))
        })
    }
}

export const removeExpense=({id}={})=>{
    return{
        type:'REMOVE_EXPENSE',
        id:id
    }
}
export const editExpense=(id,updates)=>{
    return{
        type:'EDIT_EXPENSE',
        id:id,
        updates
    }
}
// export const startEditExpense=(id,formData)=>{
//     const updates={amount,description,note,createdAt}
//     return(dispatch,getState)=>{
//         const uid=getState().authen.uid
//         return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(()=>{
//             dispatch(editExpense(id,updates))
//         })
//     }
// }
export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
      const uid = getState().authen.uid;
      return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
        dispatch(editExpense(id, updates));
      });
    };
  };

export const setExpenses=(expenses)=>{
    return{
        type:'SET_EXPENSES',
        expenses
    }
}
export const startSetExpenses=()=>{
    return(dispatch,getState)=>{
        const uid=getState().authen.uid
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot)=>{
            const expenses=[]
            snapshot.forEach((childSnapshot)=>{
                expenses.push({id:childSnapshot.key,...childSnapshot.val()})
            })
            dispatch(setExpenses(expenses))
        })
    }
}
