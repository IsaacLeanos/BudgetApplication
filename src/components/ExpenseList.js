import React from 'react'
import{connect}from'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'
import ConnectedExpenseListFilters from './ExpenseListFilters';


export const ExpenseList=(props)=>{
    return(
        <div className= "content-container">
                    <ConnectedExpenseListFilters/>
            <div className="list-header">
                <div className="show-for-mobile">Expenses</div>
                <div className="show-for-desktop">Expense</div>
                <div className="show-for-desktop">Amount</div>
            </div>
            <div className="list-body">
            {props.expenses.length===0?<div className="list-item list-item--message"> <span> No Expenses </span> </div>:
            props.expenses.map((item)=>{
            return <ExpenseListItem key={item.id} {...item}/>
        })
        }
            </div>
        
        </div>
    )
}

const mapStateToProps=(state)=>{
    return{
        expenses:selectExpenses(state.expenses,state.filters)
    }
}
export default connect(mapStateToProps)(ExpenseList)