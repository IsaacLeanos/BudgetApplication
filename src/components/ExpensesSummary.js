import React from 'react'
import{connect}from'react-redux'
import numeral from 'numeral'
import{Link}from'react-router-dom'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'



export const ExpensesSummary=({expenseCount,expensesTotal})=>{
    const formattedNumber=numeral(expensesTotal/100).format('$0,0.00')
    const word=expenseCount===1?'expense':'expenses'
    return(
        <div className='page-header'>
        <div className='content-container'>
        <h1 className="page-header__title">Viewing <span>{expenseCount}</span> {word} totaling <span>{formattedNumber}</span>
        </h1>
        <div className='page-header__actions'>
            <Link className='button' to='/create'>Add Expense</Link>
        </div>
        </div>
        </div>
    )
}

const mapStateToProps=(state)=>{
    const visibleExpenses=selectExpenses(state.expenses,state.filters)
    return{
        expenseCount:visibleExpenses.length,
        expensesTotal:selectExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)