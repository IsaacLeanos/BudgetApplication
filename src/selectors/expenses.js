import moment from 'moment'


const selectExpenses=(expenses,{text,sortBy,startDate,endDate})=>{
    return expenses.filter((item)=>{
        let createdAtMoment=moment(item.createdAt)
        let startDateMatch=startDate?startDate.isSameOrBefore(createdAtMoment,'day'):true
        let endDateMatch=endDate?endDate.isSameOrAfter(createdAtMoment,'day'):true
        let textMatch=item.description.toLowerCase().includes(text.toLowerCase())
        return startDateMatch&&endDateMatch&&textMatch
    }).sort((a,b)=>{
        if(sortBy==='date'){
            return a.createdAt<b.createdAt ? -1:1
        }else if(sortBy==='amount'){
            return a.amount<b.amount ? 1:-1
        }
    })
}

export default selectExpenses