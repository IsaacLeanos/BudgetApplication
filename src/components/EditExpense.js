import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {startEditExpense,startRemoveExpense} from '../actions/expenses'

export class EditExpensePage extends React.Component{

    onClick=()=>{
        this.props.remove({id:this.props.expense.id})
        this.props.history.push('/')
    }
    onSubmit=(formData)=>{
        this.props.startEditExpense(this.props.expense.id,formData)
        this.props.history.push('/')
    }
    render(){
        return(
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Edit Expense</h1>
                </div>
            </div>
            <div className="content-container">
            <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>
            <button className="button button--secondary" onClick={this.onClick}>Remove</button>
            </div>
        </div>
    )}
}
const mapStateToProps=(state,props)=>({
        expense:state.expenses.find((item)=>item.id===props.match.params.id)
    })

const mapDispatchToProps=(dispatch)=>({
    remove:(item)=>dispatch(startRemoveExpense(item)),
    startEditExpense:(id,item)=>dispatch(startEditExpense(id,item))
})

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)