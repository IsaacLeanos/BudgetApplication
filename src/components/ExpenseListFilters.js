import React from 'react'
import {connect} from 'react-redux'
import {setTextFilter,sortByAmount,sortByDate,setStartDate,setEndDate} from '../actions/filters'
import {DateRangePicker} from 'react-dates'
import moment from 'moment'

export class ExpenseListFilters extends React.Component{
    state={
        calendarFocused:null
    }
    onDatesChange=({startDate,endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange=(calendarFocus)=>{
        this.setState(()=>({calendarFocused:calendarFocus}))
    } 
    onInputChange=(e)=>{
        this.props.setTextFilter(e.target.value)
    }
    onSelectChange=(e)=>{
        if(e.target.value==='date'){
            this.props.selectByDate()
        }else if(e.target.value==='amount'){
            this.props.selectByAmount()
        }
    }
    render(){
        return(
        <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input type="text" className="text-input" placeholder="search expenses" value={this.props.filters.text} onChange={this.onInputChange}/>
                    </div>
                    <div className="input-group__item">
                        <select className="select" value={this.props.filters.sortBy} onChange={this.onSelectChange}>
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker startDate={this.props.filters.startDate} endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                        showClearDates={true}
                        />
                    </div>

                </div>

        </div> 

            
            
        )
    }
}
const mapStateToProps=(state)=>({filters:state.filters})

const mapDispatchToProps=(dispatch)=>({
    setStartDate:(date)=>dispatch(setStartDate(date)),
    setEndDate:(date)=>dispatch(setEndDate(date)),
    setTextFilter:(value)=>dispatch(setTextFilter(value)),
    selectByDate:()=>dispatch(sortByDate()),
    selectByAmount:()=>dispatch(sortByAmount())
})


export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters)
