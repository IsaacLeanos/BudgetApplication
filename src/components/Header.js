import React from 'react'
import{Link,NavLink}from'react-router-dom'
import{startLogout}from'../actions/authen'
import{connect}from'react-redux'


export const Header=({startLogout})=>{
    return(
        <header className="header">
        <div className="content-container">
        <div className="header__content">
        <Link className="header__title" to="/dashboard">
        <h1>Expensify</h1>
        </Link>
        <button className="button button--link" onClick={startLogout}>Logout</button>
        </div>
        </div>
        </header>
    )
}

const mapDispatchToProps=(dispatch)=>{
    return{
        startLogout:()=>{dispatch(startLogout())}
    }
}

export default connect(undefined,mapDispatchToProps)(Header)