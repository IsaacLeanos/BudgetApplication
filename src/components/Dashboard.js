import React from 'react'
import ConnectedExpenseList from './ExpenseList'
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashboardPage=()=>{
    return(
        <div>
        <ExpensesSummary/>
        <ConnectedExpenseList/>
        </div>
    )
}


export default ExpenseDashboardPage