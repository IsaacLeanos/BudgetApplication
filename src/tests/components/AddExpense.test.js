import React from 'react'
import {shallow} from 'enzyme'
import {AddExpensePage} from '../../components/AddExpense'
import expenses from '../fixtures/expensesData'

test('render AddExpensePage',()=>{
    let startAddExpense=jest.fn()
    let history={push:jest.fn()}
    let wrapper=shallow(<AddExpensePage startAddExpense={startAddExpense} history={history}/>)
    expect(wrapper).toMatchSnapshot()
})

test('handle onSubmit',()=>{
    let startAddExpense=jest.fn()
    let history={push:jest.fn()}
    let wrapper=shallow(<AddExpensePage startAddExpense={startAddExpense} history={history}/>)
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
    expect(history.push).toHaveBeenCalledWith('/')
    expect(startAddExpense).toHaveBeenCalledWith(expenses[1])
})