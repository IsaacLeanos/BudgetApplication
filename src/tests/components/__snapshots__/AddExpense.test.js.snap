// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render AddExpensePage 1`] = `
<div>
  <h1>
    Add Expense Page
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
